File: AddIntToString.scala

def addIntToString(i: Int, s: String): String = {
  def last = (carry: Int) => if (carry == 0) "" else carry.toString
  def step(f: Int => String, i: Int) = (x: Int) => {
    val y = (x % 10) + i
    f((y / 10) + (x / 10)) + (y % 10).toString
  }
  s.toList.map(_.toInt - '0').foldLeft(last)(step)(i)
}

println(addIntToString(1, "1999999999999999999999999999999"))
println(addIntToString(2, "0000000000000000009999999999999"))
println(addIntToString(3, "0000000000000000000000000000000"))
println(addIntToString(123456789, "0"))

File: AddIntToString.py
def addCont(s, i): 
    def last(carry):
        if carry != 0:
            return str(carry)
        else:
            return ""
    def step(k, x):
        def f(carry):
            y      = int(x) + carry % 10
            carry2 = carry / 10 + y / 10
            x2     = str(y % 10) 
            return k(carry2) + x2
        return f
    return reduce(step, s, last)(i)

print addCont("234172387421873412341", 1234)
print addCont("99", 10000)
